File: src/cloudtwin/providers/azure/servicebus/handlers.py
"""
Azure Service Bus HTTP handlers – JSON REST API.

Since the azure-servicebus SDK uses AMQP (not HTTP) for message operations,
this implements a custom JSON REST API that can be tested directly with httpx:

  Queues:
    PUT    /{ns}/queues/{queue}                        → create queue
    GET    /{ns}/queues                                → list queues
    GET    /{ns}/queues/{queue}                        → get queue
    DELETE /{ns}/queues/{queue}                        → delete queue
    POST   /{ns}/queues/{queue}/messages               → send message
    GET    /{ns}/queues/{queue}/messages               → receive messages (?limit=)
    DELETE /{ns}/queues/{queue}/messages/{lock_token}  → complete/delete message
    POST   /{ns}/queues/{queue}/messages/{lock_token}/abandon     → abandon
    POST   /{ns}/queues/{queue}/messages/{lock_token}/deadletter  → dead-letter

  Topics + Subscriptions:
    PUT    /{ns}/topics/{topic}                                  → create topic
    GET    /{ns}/topics                                          → list topics
    DELETE /{ns}/topics/{topic}                                  → delete topic
    PUT    /{ns}/topics/{topic}/subscriptions/{sub}             → create subscription
    GET    /{ns}/topics/{topic}/subscriptions                   → list subscriptions
    POST   /{ns}/topics/{topic}/messages                        → publish to topic
    GET    /{ns}/topics/{topic}/subscriptions/{sub}/messages    → receive from subscription
"""

from __future__ import annotations

import dataclasses
import logging
from typing import Optional

from fastapi import APIRouter, Query, Request
from fastapi.responses import JSONResponse

from cloudtwin.core.errors import NotFoundError
from cloudtwin.providers.azure.servicebus.service import ServiceBusService

log = logging.getLogger("cloudtwin.azure.servicebus")


def _msg_dict(msg) -> dict:
    return {
        "message_id": msg.message_id,
        "lock_token": msg.lock_token,
        "body": msg.body,
        "content_type": msg.content_type,
        "state": msg.state,
        "delivery_count": msg.delivery_count,
        "created_at": msg.created_at,
    }


def make_router(service: ServiceBusService, namespace: str) -> APIRouter:
    router = APIRouter()

    # ------------------------------------------------------------------
    # Queues
    # ------------------------------------------------------------------

    @router.put(f"/{namespace}/queues/{{queue_name}}")
    async def create_queue(queue_name: str) -> JSONResponse:
        queue = await service.create_queue(queue_name)
        return JSONResponse(
            {
                "name": queue.name,
                "namespace": queue.namespace,
                "created_at": queue.created_at,
            },
            status_code=201,
        )

    @router.get(f"/{namespace}/queues")
    async def list_queues() -> JSONResponse:
        queues = await service.list_queues()
        return JSONResponse(
            {"queues": [{"name": q.name, "created_at": q.created_at} for q in queues]}
        )

    @router.get(f"/{namespace}/queues/{{queue_name}}")
    async def get_queue(queue_name: str) -> JSONResponse:
        try:
            queue = await service.get_queue(queue_name)
        except NotFoundError as exc:
            return JSONResponse({"error": exc.message}, status_code=404)
        return JSONResponse(
            {
                "name": queue.name,
                "namespace": queue.namespace,
                "created_at": queue.created_at,
            }
        )

    @router.delete(f"/{namespace}/queues/{{queue_name}}")
    async def delete_queue(queue_name: str) -> JSONResponse:
        try:
            await service.delete_queue(queue_name)
        except NotFoundError as exc:
            return JSONResponse({"error": exc.message}, status_code=404)
        return JSONResponse({}, status_code=204)

    @router.post(f"/{namespace}/queues/{{queue_name}}/messages")
    async def send_queue_message(queue_name: str, request: Request) -> JSONResponse:
        body_bytes = await request.body()
        content_type = request.headers.get("content-type", "text/plain")
        try:
            msg = await service.send_to_queue(
                queue_name, body_bytes.decode(), content_type
            )
        except NotFoundError as exc:
            return JSONResponse({"error": exc.message}, status_code=404)
        return JSONResponse(_msg_dict(msg), status_code=201)

    @router.get(f"/{namespace}/queues/{{queue_name}}/messages")
    async def receive_queue_messages(
        queue_name: str, limit: int = Query(default=1)
    ) -> JSONResponse:
        try:
            messages = await service.receive_from_queue(queue_name, limit=limit)
        except NotFoundError as exc:
            return JSONResponse({"error": exc.message}, status_code=404)
        return JSONResponse({"messages": [_msg_dict(m) for m in messages]})

    @router.delete(f"/{namespace}/queues/{{queue_name}}/messages/{{lock_token}}")
    async def complete_queue_message(queue_name: str, lock_token: str) -> JSONResponse:
        try:
            await service.complete_message(lock_token)
        except NotFoundError as exc:
            return JSONResponse({"error": exc.message}, status_code=404)
        return JSONResponse({}, status_code=204)

    @router.post(f"/{namespace}/queues/{{queue_name}}/messages/{{lock_token}}/abandon")
    async def abandon_queue_message(queue_name: str, lock_token: str) -> JSONResponse:
        try:
            await service.abandon_message(lock_token)
        except NotFoundError as exc:
            return JSONResponse({"error": exc.message}, status_code=404)
        return JSONResponse({})

    @router.post(
        f"/{namespace}/queues/{{queue_name}}/messages/{{lock_token}}/deadletter"
    )
    async def deadletter_queue_message(
        queue_name: str, lock_token: str
    ) -> JSONResponse:
        try:
            await service.deadletter_message(lock_token)
        except NotFoundError as exc:
            return JSONResponse({"error": exc.message}, status_code=404)
        return JSONResponse({})

    # ------------------------------------------------------------------
    # Topics
    # ------------------------------------------------------------------

    @router.put(f"/{namespace}/topics/{{topic_name}}")
    async def create_topic(topic_name: str) -> JSONResponse:
        topic = await service.create_topic(topic_name)
        return JSONResponse(
            {
                "name": topic.name,
                "namespace": topic.namespace,
                "created_at": topic.created_at,
            },
            status_code=201,
        )

    @router.get(f"/{namespace}/topics")
    async def list_topics() -> JSONResponse:
        topics = await service.list_topics()
        return JSONResponse(
            {"topics": [{"name": t.name, "created_at": t.created_at} for t in topics]}
        )

    @router.delete(f"/{namespace}/topics/{{topic_name}}")
    async def delete_topic(topic_name: str) -> JSONResponse:
        try:
            await service.delete_topic(topic_name)
        except NotFoundError as exc:
            return JSONResponse({"error": exc.message}, status_code=404)
        return JSONResponse({}, status_code=204)

    # ------------------------------------------------------------------
    # Subscriptions
    # ------------------------------------------------------------------

    @router.put(f"/{namespace}/topics/{{topic_name}}/subscriptions/{{sub_name}}")
    async def create_subscription(topic_name: str, sub_name: str) -> JSONResponse:
        try:
            sub = await service.create_subscription(topic_name, sub_name)
        except NotFoundError as exc:
            return JSONResponse({"error": exc.message}, status_code=404)
        return JSONResponse(
            {"name": sub.name, "topic_id": sub.topic_id, "created_at": sub.created_at},
            status_code=201,
        )

    @router.get(f"/{namespace}/topics/{{topic_name}}/subscriptions")
    async def list_subscriptions(topic_name: str) -> JSONResponse:
        try:
            subs = await service.list_subscriptions(topic_name)
        except NotFoundError as exc:
            return JSONResponse({"error": exc.message}, status_code=404)
        return JSONResponse(
            {
                "subscriptions": [
                    {"name": s.name, "created_at": s.created_at} for s in subs
                ]
            }
        )

    @router.post(f"/{namespace}/topics/{{topic_name}}/messages")
    async def publish_to_topic(topic_name: str, request: Request) -> JSONResponse:
        body_bytes = await request.body()
        content_type = request.headers.get("content-type", "text/plain")
        try:
            msgs = await service.send_to_topic(
                topic_name, body_bytes.decode(), content_type
            )
        except NotFoundError as exc:
            return JSONResponse({"error": exc.message}, status_code=404)
        return JSONResponse({"fan_out": len(msgs)}, status_code=201)

    @router.get(
        f"/{namespace}/topics/{{topic_name}}/subscriptions/{{sub_name}}/messages"
    )
    async def receive_from_subscription(
        topic_name: str, sub_name: str, limit: int = Query(default=1)
    ) -> JSONResponse:
        try:
            messages = await service.receive_from_subscription(
                topic_name, sub_name, limit=limit
            )
        except NotFoundError as exc:
            return JSONResponse({"error": exc.message}, status_code=404)
        return JSONResponse({"messages": [_msg_dict(m) for m in messages]})

    @router.delete(
        f"/{namespace}/topics/{{topic_name}}/subscriptions/{{sub_name}}/messages/{{lock_token}}"
    )
    async def complete_sub_message(
        topic_name: str, sub_name: str, lock_token: str
    ) -> JSONResponse:
        try:
            await service.complete_message(lock_token)
        except NotFoundError as exc:
            return JSONResponse({"error": exc.message}, status_code=404)
        return JSONResponse({}, status_code=204)

    return router

File: rust/cloudtwin-lite/src/azure/servicebus/handlers.rs
//! Azure Service Bus HTTP handlers.

use std::sync::Arc;

use axum::{
    body::Bytes,
    extract::{Path, State},
    http::StatusCode,
    response::{IntoResponse, Response},
    routing::{delete, get, post, put},
    Json, Router,
};

use super::service::ServiceBusService;
use crate::telemetry;
use crate::AppState;

fn svc(state: &Arc<AppState>) -> ServiceBusService {
    ServiceBusService::new(state.db.clone(), state.cfg.azure_namespace.clone())
}

fn err(status: StatusCode, msg: &str) -> Response {
    (status, Json(serde_json::json!({ "error": msg }))).into_response()
}

/// Router is merged into the /azure prefix in main.
pub fn router() -> Router<Arc<AppState>> {
    Router::new()
        // Queues
        .route("/:namespace/queues", get(list_queues))
        .route(
            "/:namespace/queues/:queue",
            put(create_queue).delete(delete_queue),
        )
        .route(
            "/:namespace/queues/:queue/messages",
            post(send_message).get(receive_messages),
        )
        .route(
            "/:namespace/queues/:queue/messages/:lock_token",
            delete(complete_message),
        )
        .route(
            "/:namespace/queues/:queue/messages/:lock_token/abandon",
            post(abandon_message),
        )
        // Topics
        .route("/:namespace/topics", get(list_topics))
        .route(
            "/:namespace/topics/:topic",
            put(create_topic).delete(delete_topic),
        )
        .route(
            "/:namespace/topics/:topic/messages",
            post(send_topic_message),
        )
        // Subscriptions
        .route(
            "/:namespace/topics/:topic/subscriptions",
            get(list_subscriptions),
        )
        .route(
            "/:namespace/topics/:topic/subscriptions/:sub",
            put(create_subscription),
        )
        .route(
            "/:namespace/topics/:topic/subscriptions/:sub/messages",
            get(receive_sub_messages),
        )
        .route(
            "/:namespace/topics/:topic/subscriptions/:sub/messages/:lock_token",
            delete(complete_sub_message),
        )
}

// ── Queue handlers ────────────────────────────────────────────────────────────

async fn create_queue(
    State(state): State<Arc<AppState>>,
    Path((_, queue)): Path<(String, String)>,
) -> Response {
    match svc(&state).create_queue(&queue).await {
        Ok(q) => {
            telemetry::emit(&state.db, "azure", "servicebus", "create_queue", &serde_json::json!({"queue": queue}).to_string()).await;
            (StatusCode::CREATED, Json(serde_json::json!({ "name": q.name }))).into_response()
        }
        Err(e) => err(StatusCode::INTERNAL_SERVER_ERROR, &e.to_string()),
    }
}

async fn list_queues(State(state): State<Arc<AppState>>, Path(_ns): Path<String>) -> Response {
    match svc(&state).list_queues().await {
        Ok(qs) => {
            let items: Vec<_> = qs
                .iter()
                .map(|q| serde_json::json!({ "name": q.name }))
                .collect();
            (StatusCode::OK, Json(serde_json::json!({ "value": items }))).into_response()
        }
        Err(e) => err(StatusCode::INTERNAL_SERVER_ERROR, &e.to_string()),
    }
}

async fn delete_queue(
    State(state): State<Arc<AppState>>,
    Path((_, queue)): Path<(String, String)>,
) -> StatusCode {
    match svc(&state).delete_queue(&queue).await {
        Ok(_) => {
            let _ = telemetry::emit(&state.db, "azure", "servicebus", "delete_queue", &serde_json::json!({"queue": queue}).to_string()).await;
            StatusCode::OK
        }
        Err(_) => StatusCode::INTERNAL_SERVER_ERROR,
    }
}

async fn send_message(
    State(state): State<Arc<AppState>>,
    Path((_, queue)): Path<(String, String)>,
    body: Bytes,
) -> Response {
    let body_str = String::from_utf8_lossy(&body).to_string();
    match svc(&state).send_message(&queue, &body_str).await {
        Ok(mid) => {
            telemetry::emit(&state.db, "azure", "servicebus", "send_message", &serde_json::json!({"queue": queue}).to_string()).await;
            (StatusCode::CREATED, Json(serde_json::json!({ "messageId": mid }))).into_response()
        }
        Err(e) => err(StatusCode::INTERNAL_SERVER_ERROR, &e.to_string()),
    }
}

async fn receive_messages(
    State(state): State<Arc<AppState>>,
    Path((_, queue)): Path<(String, String)>,
) -> Response {
    match svc(&state).receive_messages(&queue, 1).await {
        Ok(msgs) => {
            let items: Vec<_> = msgs
                .iter()
                .map(|m| {
                    serde_json::json!({
                        "messageId":  m.message_id,
                        "body":       m.body,
                        "lockToken":  m.lock_token,
                    })
                })
                .collect();
            (StatusCode::OK, Json(serde_json::json!({ "value": items }))).into_response()
        }
        Err(e) => err(StatusCode::INTERNAL_SERVER_ERROR, &e.to_string()),
    }
}

async fn complete_message(
    State(state): State<Arc<AppState>>,
    Path((_, _, lock_token)): Path<(String, String, String)>,
) -> StatusCode {
    match svc(&state).complete_message(&lock_token).await {
        Ok(_) => StatusCode::OK,
        Err(_) => StatusCode::INTERNAL_SERVER_ERROR,
    }
}

async fn abandon_message(
    State(state): State<Arc<AppState>>,
    Path((_, _, lock_token)): Path<(String, String, String)>,
) -> StatusCode {
    match svc(&state).abandon_message(&lock_token).await {
        Ok(_) => StatusCode::OK,
        Err(_) => StatusCode::INTERNAL_SERVER_ERROR,
    }
}

// ── Topic handlers ────────────────────────────────────────────────────────────

async fn create_topic(
    State(state): State<Arc<AppState>>,
    Path((_, topic)): Path<(String, String)>,
) -> Response {
    match svc(&state).create_topic(&topic).await {
        Ok(t) => {
            telemetry::emit(&state.db, "azure", "servicebus", "create_topic", &serde_json::json!({"topic": topic}).to_string()).await;
            (StatusCode::CREATED, Json(serde_json::json!({ "name": t.name }))).into_response()
        }
        Err(e) => err(StatusCode::INTERNAL_SERVER_ERROR, &e.to_string()),
    }
}

async fn list_topics(State(state): State<Arc<AppState>>, Path(_ns): Path<String>) -> Response {
    match svc(&state).list_topics().await {
        Ok(ts) => {
            let items: Vec<_> = ts
                .iter()
                .map(|t| serde_json::json!({ "name": t.name }))
                .collect();
            (StatusCode::OK, Json(serde_json::json!({ "value": items }))).into_response()
        }
        Err(e) => err(StatusCode::INTERNAL_SERVER_ERROR, &e.to_string()),
    }
}

async fn delete_topic(
    State(state): State<Arc<AppState>>,
    Path((_, topic)): Path<(String, String)>,
) -> StatusCode {
    match svc(&state).delete_topic(&topic).await {
        Ok(_) => StatusCode::OK,
        Err(_) => StatusCode::INTERNAL_SERVER_ERROR,
    }
}

async fn send_topic_message(
    State(state): State<Arc<AppState>>,
    Path((_, topic)): Path<(String, String)>,
    body: Bytes,
) -> Response {
    let body_str = String::from_utf8_lossy(&body).to_string();
    match svc(&state).send_topic_message(&topic, &body_str).await {
        Ok(mid) => (
            StatusCode::CREATED,
            Json(serde_json::json!({ "messageId": mid })),
        )
            .into_response(),
        Err(e) => err(StatusCode::INTERNAL_SERVER_ERROR, &e.to_string()),
    }
}

// ── Subscription handlers ─────────────────────────────────────────────────────

async fn create_subscription(
    State(state): State<Arc<AppState>>,
    Path((_, topic, sub)): Path<(String, String, String)>,
) -> Response {
    match svc(&state).create_subscription(&topic, &sub).await {
        Ok(s) => {
            telemetry::emit(&state.db, "azure", "servicebus", "create_subscription", &serde_json::json!({"topic": topic, "subscription": sub}).to_string()).await;
            (StatusCode::CREATED, Json(serde_json::json!({ "name": s.name, "topic": s.topic }))).into_response()
        }
        Err(e) => err(StatusCode::INTERNAL_SERVER_ERROR, &e.to_string()),
    }
}

async fn list_subscriptions(
    State(state): State<Arc<AppState>>,
    Path((_, topic)): Path<(String, String)>,
) -> Response {
    match svc(&state).list_subscriptions(&topic).await {
        Ok(subs) => {
            let items: Vec<_> = subs
                .iter()
                .map(|s| serde_json::json!({ "name": s.name }))
                .collect();
            (StatusCode::OK, Json(serde_json::json!({ "value": items }))).into_response()
        }
        Err(e) => err(StatusCode::INTERNAL_SERVER_ERROR, &e.to_string()),
    }
}

async fn receive_sub_messages(
    State(state): State<Arc<AppState>>,
    Path((_, topic, sub)): Path<(String, String, String)>,
) -> Response {
    match svc(&state)
        .receive_subscription_messages(&topic, &sub, 1)
        .await
    {
        Ok(msgs) => {
            let items: Vec<_> = msgs
                .iter()
                .map(|m| {
                    serde_json::json!({
                        "messageId": m.message_id,
                        "body":      m.body,
                        "lockToken": m.lock_token,
                    })
                })
                .collect();
            (StatusCode::OK, Json(serde_json::json!({ "value": items }))).into_response()
        }
        Err(e) => err(StatusCode::INTERNAL_SERVER_ERROR, &e.to_string()),
    }
}

async fn complete_sub_message(
    State(state): State<Arc<AppState>>,
    Path((_, _, _, lock_token)): Path<(String, String, String, String)>,
) -> StatusCode {
    match svc(&state).complete_message(&lock_token).await {
        Ok(_) => StatusCode::OK,
        Err(_) => StatusCode::INTERNAL_SERVER_ERROR,
    }
}
